drivers/led_strip/ws2812_spi.c

```diff
@@ -10,54 +11,68 @@
 
 #define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
 #include <logging/log.h>
-LOG_MODULE_REGISTER(ws2812);
+LOG_MODULE_REGISTER(ws2812_spi);
 
 #include <zephyr.h>
 #include <device.h>
 #include <drivers/spi.h>
 #include <sys/util.h>
 
+#define DEV_LABEL    DT_INST_0_WORLDSEMI_WS2812_SPI_LABEL
+
 /*
  * WS2812-ish SPI master configuration:
  *
  * - mode 0 (the default), 8 bit, MSB first (arbitrary), one-line SPI
  * - no shenanigans (don't hold CS, don't hold the device lock, this
  *   isn't an EEPROM)
  */
-#define SPI_OPER (SPI_OP_MODE_MASTER |		\
-		  SPI_TRANSFER_MSB |		\
-		  SPI_WORD_SET(8) |		\
-		  SPI_LINES_SINGLE)
-
-#define SPI_FREQ              DT_INST_0_WORLDSEMI_WS2812_SPI_MAX_FREQUENCY
-#define ONE_FRAME             CONFIG_WS2812_STRIP_ONE_FRAME
-#define ZERO_FRAME            CONFIG_WS2812_STRIP_ZERO_FRAME
-#define RED_OFFSET            (8 * sizeof(u8_t) * CONFIG_WS2812_RED_ORDER)
-#define GRN_OFFSET            (8 * sizeof(u8_t) * CONFIG_WS2812_GRN_ORDER)
-#define BLU_OFFSET            (8 * sizeof(u8_t) * CONFIG_WS2812_BLU_ORDER)
-#ifdef CONFIG_WS2812_HAS_WHITE_CHANNEL
-#define WHT_OFFSET            (8 * sizeof(u8_t) * CONFIG_WS2812_WHT_ORDER)
+#define SPI_BUS      DT_INST_0_WORLDSEMI_WS2812_SPI_BUS_NAME
+#define SPI_SLAVE    DT_INST_0_WORLDSEMI_WS2812_SPI_BASE_ADDRESS
+#define SPI_OPER     (SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | \
+		      SPI_WORD_SET(8) | SPI_LINES_SINGLE)
+#define SPI_FREQ     DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_MAX_FREQUENCY
+#define ONE_FRAME    DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_ONE_FRAME
+#define ZERO_FRAME   DT_INST_0_WORLDSEMI_WS2812_SPI_SPI_ZERO_FRAME
+
+#define RED_ORDER    DT_INST_0_WORLDSEMI_WS2812_SPI_RED_ORDER
+#define GREEN_ORDER  DT_INST_0_WORLDSEMI_WS2812_SPI_GREEN_ORDER
+#define BLUE_ORDER   DT_INST_0_WORLDSEMI_WS2812_SPI_BLUE_ORDER
+#define WHITE_ORDER  DT_INST_0_WORLDSEMI_WS2812_SPI_WHITE_ORDER
```

Comment:
We probably should create `struct ws2812_config` and place all these data there. It will make it easier to extend the driver in the future by adding extra instances. The same could be done for `ws2812_gpio.c`.

Reply:
I'm trying to keep the loops operating on build time data rather than run-time data, so the compiler can optimize better.

I'd actually rather delete support for this whole feature than add flash and run-time overhead for it. I've seen only GRB or GRBW; I threw this in just because I heard there was some variation possible somewhere on the internet :).

What do you think?

Reply:
I believe I haven't explained myself very well. This is meant to be a fully fledged driver, even if we start support with one specific target. I'm concern about the situation where user wants to connect more than one WS2813 led strip to their board (having two or more SPI peripherals or using different set of GPIO pins in case of bit-bang driver). This is handled by creating multiple instances of the driver. In this case all the instance specific data need to go into the config struct available to the driver via `dev->config->config_info`. So my comment concerns all `DT_INST_0_` defines (I used multi-line comment but they don't stand out very well).

It's true that in this case the amount of optimizations the compiler can perform is smaller.

Reply:
> This is meant to be a fully fledged driver, even if we start support with one specific target.

It would be nice if drivers generally supported multiple instances, but Zephyr is a long way from reaching that goal, and there are cases like this where it's simply not reasonable.  I'm open to a consensus decision that this driver be documented as supporting only a single device instance.

If it consensus is unreachable I would like to see this converted to an example that's modularized enough to allow it to be copied into applications that need it.  It doesn't sound like this particular implementation is suited to being a "fully fledged driver" even though it uses the API and is certainly valuable.

Reply:
> m concern about the situation where user wants to connect more than one WS2813 led strip to their board (having two or more SPI peripherals or using different set of GPIO pins in case of bit-bang driver). This is handled by creating multiple instances of the driver.

Yup, I understand what you're saying.

What I'm saying is that if we're going to make this driver multi-instance, I would rather delete the order configuration support since I think it's likely to be rarely needed, than inflate the flash and runtime requirements of the driver. Either way is fine with me.

Reply:
> if we're going to make this driver multi-instance (even hypothetically in the future), I would rather delete the order configuration support since I think it's likely to be rarely needed

Please treat my comment regarding multiple instances as a suggestion. I'm fine with developing this driver in steps. That said, if we don't plan to support the order configuration long term it might be better to drop it now. I do believe that at some time point we should add support for multiple instances. On the other hand "order configuration" could be made a conditional code enabled via Kconfig option. That would also be a viable approach so no need to hurry here with the decisions either.